File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/CBuilder/DragDropDemo/DragDropDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DragDropDemoMain.h"
#include "DragDropDemoData.h"
#include "DragDropDemoDictionary.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxCheckBox"
#pragma link "cxCurrencyEdit"
#pragma link "cxDBTL"
#pragma link "cxInplaceContainer"
#pragma link "cxMaskEdit"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma resource "*.dfm"
TDragDropDemoMainForm *DragDropDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TDragDropDemoMainForm::TDragDropDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::FormShow(TObject *Sender)
{
  tlDepartments->FullExpand();
  DragDropDemoDictionaryForm->Show();

/* remove/add the closing slash on this line to disable/enable the following code *

  ShowMessage("WARNING: tutorial not completed. First, please apply the steps "
    "shown in the doc file");

//*/
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::miColumnCustomizationClick(TObject *Sender)
{
  tlDepartments->Customizing->Visible = true;
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::miDragCollapseClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  tlDepartments->OptionsBehavior->DragCollapse = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::miDragExpandeClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  tlDepartments->OptionsBehavior->DragExpand = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::ShowDictionaries1Click(TObject *Sender)
{
  DragDropDemoDictionaryForm->Show();
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::tlDepartmentsInitInsertingRecord(
  TcxCustomDBTreeList *Sender, TcxDBTreeListNode *AFocusedNode, bool &AHandled)
{
  if (AFocusedNode != NULL)
    DragDropDemoDataDM->SetParentValue(AFocusedNode->ParentKeyValue);
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::tlDepartmentsDragDrop(TObject *Sender,
  TObject *Source, int X,  int Y)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  if (Sender == Source || !IsHitAtNode((TcxDBTreeList*)Sender, X, Y))
    return;
  TcxDBTreeListNode *AHitNode =
    (TcxDBTreeListNode*)((TcxDBTreeList*)Sender)->HitTest->HitNode;
  if (Source == DragDropDemoDictionaryForm->tlDeptDict){
    if (IsDropAsChild((TcxDBTreeList*)Sender))
      DragDropDemoDictionaryForm->SetDeptSelectionParentValue(AHitNode->KeyValue);
    else
      DragDropDemoDictionaryForm->SetDeptSelectionParentValue(AHitNode->ParentKeyValue);
    ((TcxDBTreeList*)Sender)->DataController->DataSet->Refresh();
  }
  else{
    if (Source == DragDropDemoDictionaryForm->tlEmplDict)
      SetEmplDictSelectionDeptID(AHitNode->KeyValue);
    else
      SetSelectedNodesValue(tlEmployees, tlEmployeesDepartmentID->ItemIndex,
        AHitNode->KeyValue);
  }

//*/
}

bool TDragDropDemoMainForm::IsDropAsChild(TcxDBTreeList *Sender)
{
   return !(Sender->HitTest->HitAtIndent || Sender->HitTest->HitAtIndicator);
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::tlDepartmentsDragOver(TObject *Sender,
  TObject *Source, int X, int Y, TDragState State, bool &Accept)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  Accept = IsHitAtNode((TcxDBTreeList*)Sender, X, Y);

//*/
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::tlEmployeesDragDrop(TObject *Sender,
  TObject *Source, int X,  int Y)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  if (Source == DragDropDemoDictionaryForm->tlEmplDict){
    Variant AValue;
    if (tlDepartments->FocusedNode != NULL)
      AValue = ((TcxDBTreeListNode*)tlDepartments->FocusedNode)->KeyValue;
    else
      AValue = -1;
    SetEmplDictSelectionDeptID(AValue);
  }

//*/
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::tlEmployeesDragOver(TObject *Sender,
  TObject *Source, int X, int Y, TDragState State, bool &Accept)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  Accept = (Source == DragDropDemoDictionaryForm->tlEmplDict);

//*/
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoMainForm::tlEmployeesMoveTo(TcxCustomTreeList *Sender,
  TcxTreeListNode *AttachNode, TcxTreeListNodeAttachMode AttachMode,
  TList Nodes, bool &IsCopy, bool &Done)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  Done = true;

//*/
}
//---------------------------------------------------------------------------

bool TDragDropDemoMainForm::IsHitAtNode(TcxDBTreeList *ATreeList, int X, int Y)
{
    ATreeList->HitTest->ReCalculate(Point(X,Y));
    return ATreeList->HitTest->HitAtNode;
}
//---------------------------------------------------------------------------

void TDragDropDemoMainForm::SetEmplDictSelectionDeptID(Variant AValue)
{
  DragDropDemoDictionaryForm->SetEmplSelectionDeptID(AValue);
  tlEmployees->DataController->DataSet->Close();
  tlEmployees->DataController->DataSet->Open();
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/Delphi/DragDropDemo/DragDropDemoMain.pas
unit DragDropDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  cxStyles, cxTL, cxMaskEdit, cxCurrencyEdit, cxMemo, cxCheckBox, Menus,
  cxLookAndFeels, ActnList, ImgList, cxInplaceContainer, cxDBTL,
  cxControls, cxTLData, ComCtrls, StdCtrls, DemoBasicMain, ExtCtrls,
  cxGraphics, cxCustomData, Dialogs, cxLookAndFeelPainters;

type
  TcxDragSenderType = (dotDepart, dotPers, dotNone);

  TDragDropDemoMainForm = class(TDemoBasicMainForm)
    miSeparator1: TMenuItem;
    miColumnCustomization: TMenuItem;
    Splitter1: TSplitter;
    pnlDepartments: TPanel;
    tlDepartments: TcxDBTreeList;
    tlDepartmentsID: TcxDBTreeListColumn;
    tlDepartmentsPARENTID: TcxDBTreeListColumn;
    tlDepartmentsNAME: TcxDBTreeListColumn;
    tlDepartmentsBUDGET: TcxDBTreeListColumn;
    tlDepartmentsPHONE: TcxDBTreeListColumn;
    tlDepartmentsFAX: TcxDBTreeListColumn;
    tlDepartmentsEMAIL: TcxDBTreeListColumn;
    tlDepartmentsVACANCY: TcxDBTreeListColumn;
    pnlDeptCaption: TPanel;
    pnlEmployees: TPanel;
    pnlEmplCaption: TPanel;
    tlEmployees: TcxDBTreeList;
    tlEmployeesName: TcxDBTreeListColumn;
    tlEmployeesCountry: TcxDBTreeListColumn;
    tlEmployeesPostalCode: TcxDBTreeListColumn;
    tlEmployeesCity: TcxDBTreeListColumn;
    tlEmployeesAddress: TcxDBTreeListColumn;
    tlEmployeesPhone: TcxDBTreeListColumn;
    tlEmployeesFax: TcxDBTreeListColumn;
    tlEmployeesEMAIL: TcxDBTreeListColumn;
    tlEmployeesHOMEPAGE: TcxDBTreeListColumn;
    tlEmployeesDepartmentID: TcxDBTreeListColumn;
    N1: TMenuItem;
    miDragExpande: TMenuItem;
    miDragCollapse: TMenuItem;
    N2: TMenuItem;
    ShowDictionaries1: TMenuItem;
    procedure FormShow(Sender: TObject);
    procedure miColumnCustomizationClick(Sender: TObject);
    procedure miDragCollapseClick(Sender: TObject);
    procedure miDragExpandeClick(Sender: TObject);
    procedure ShowDictionaries1Click(Sender: TObject);
    procedure tlDepartmentsInitInsertingRecord(Sender: TcxCustomDBTreeList;
      AFocusedNode: TcxDBTreeListNode; var AHandled: Boolean);
    procedure tlDepartmentsDragDrop(Sender, Source: TObject; X,
      Y: Integer);
    procedure tlDepartmentsDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
    procedure tlEmployeesDragDrop(Sender, Source: TObject; X, Y: Integer);
    procedure tlEmployeesDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
    procedure tlEmployeesMoveTo(Sender: TcxCustomTreeList;
      AttachNode: TcxTreeListNode; AttachMode: TcxTreeListNodeAttachMode;
      Nodes: TList; var IsCopy, Done: Boolean);
  private
    function IsHitAtNode(ATreeList: TcxDBTreeList; X, Y: Integer): Boolean;
    procedure SetEmplDictSelectionDeptID(AValue: Variant);
  end;

var
  DragDropDemoMainForm: TDragDropDemoMainForm;

implementation

uses
  DragDropDemoData, ShellAPI, DB, DragDropDemoDictionary, SkinDemoUtils;

{$R *.dfm}

procedure TDragDropDemoMainForm.FormShow(Sender: TObject);
begin
  tlDepartments.FullExpand;
  DragDropDemoDictionaryForm.Show;

{ remove/add the closing brace on this line to disable/enable the following code

  ShowMessage('WARNING: tutorial not completed. First, please apply the steps '+
              'shown in the doc file');

//}
end;

procedure TDragDropDemoMainForm.miColumnCustomizationClick(Sender: TObject);
begin
  tlDepartments.Customizing.Visible := True;
end;

procedure TDragDropDemoMainForm.miDragCollapseClick(Sender: TObject);
begin
{$IFNDEF EXPRESSBARS}
  SetMenuItemChecked(Sender, not GetMenuItemChecked(Sender));
{$ENDIF}
  tlDepartments.OptionsBehavior.DragCollapse := GetMenuItemChecked(Sender);
end;

procedure TDragDropDemoMainForm.miDragExpandeClick(Sender: TObject);
begin
{$IFNDEF EXPRESSBARS}
  SetMenuItemChecked(Sender, not GetMenuItemChecked(Sender));
{$ENDIF}
  tlDepartments.OptionsBehavior.DragExpand := GetMenuItemChecked(Sender);
end;

procedure TDragDropDemoMainForm.ShowDictionaries1Click(Sender: TObject);
begin
  DragDropDemoDictionaryForm.Show;
end;

procedure TDragDropDemoMainForm.tlDepartmentsInitInsertingRecord(
  Sender: TcxCustomDBTreeList; AFocusedNode: TcxDBTreeListNode;
  var AHandled: Boolean);
begin
  if AFocusedNode <> nil then
    DragDropDemoDataDM.SetParentValue(AFocusedNode.ParentKeyValue);
end;

procedure TDragDropDemoMainForm.tlDepartmentsDragDrop(Sender,
  Source: TObject; X, Y: Integer);
  function IsDropAsChild: Boolean;
  begin
    with TcxDBTreeList(Sender) do
      Result := not (HitTest.HitAtIndent or HitTest.HitAtIndicator)
  end;
var
  AHitNode: TcxDBTreeListNode;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if (Sender = Source) or not IsHitAtNode(TcxDBTreeList(Sender), X, Y) then
    Exit;
  AHitNode := TcxDBTreeListNode(TcxDBTreeList(Sender).HitTest.HitNode);
  with DragDropDemoDictionaryForm do
    if Source = tlDeptDict then
    begin
      if IsDropAsChild then
        SetDeptSelectionParentValue(AHitNode.KeyValue)
      else
        SetDeptSelectionParentValue(AHitNode.ParentKeyValue);
      TcxDBTreeList(Sender).DataController.DataSet.Refresh;
    end
    else
    begin
      if Source = tlEmplDict then
        SetEmplDictSelectionDeptID(AHitNode.KeyValue)
      else
        SetSelectedNodesValue(tlEmployees, tlEmployeesDepartmentID.ItemIndex,
          AHitNode.KeyValue);
    end;

//}
end;

procedure TDragDropDemoMainForm.tlDepartmentsDragOver(Sender,
  Source: TObject; X, Y: Integer; State: TDragState; var Accept: Boolean);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Accept := IsHitAtNode(TcxDBTreeList(Sender), X, Y);

//}
end;

procedure TDragDropDemoMainForm.tlEmployeesDragDrop(Sender, Source: TObject;
  X, Y: Integer);
var
  AValue: Variant;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if Source = DragDropDemoDictionaryForm.tlEmplDict then
  begin
    if tlDepartments.FocusedNode <> nil then
      AValue := TcxDBTreeListNode(tlDepartments.FocusedNode).KeyValue
    else
      AValue := -1;
    SetEmplDictSelectionDeptID(AValue)
  end;

//}
end;

procedure TDragDropDemoMainForm.tlEmployeesDragOver(Sender, Source: TObject;
  X, Y: Integer; State: TDragState; var Accept: Boolean);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Accept := (Source = DragDropDemoDictionaryForm.tlEmplDict);

//}
end;

procedure TDragDropDemoMainForm.tlEmployeesMoveTo(Sender: TcxCustomTreeList;
  AttachNode: TcxTreeListNode; AttachMode: TcxTreeListNodeAttachMode;
  Nodes: TList; var IsCopy, Done: Boolean);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Done := True;

//}
end;

function TDragDropDemoMainForm.IsHitAtNode(ATreeList: TcxDBTreeList; X,
  Y: Integer): Boolean;
begin
  with ATreeList do
  begin
    HitTest.ReCalculate(Point(X,Y));
    Result := HitTest.HitAtNode;
  end;
end;

procedure TDragDropDemoMainForm.SetEmplDictSelectionDeptID(AValue: Variant);
begin
  DragDropDemoDictionaryForm.SetEmplSelectionDeptID(AValue);
  tlEmployees.DataController.DataSet.Close;
  tlEmployees.DataController.DataSet.Open;
end;

end.
